UI-Impl pull request

KieranRodgerson opened the pull request:
Added ImGui library to files and created UISystem class.

KieranRodgerson committed:
ImGui Implemented

ImGui has been implemented and some basic code has been written to initialise it

KieranRodgerson committed:
Window.h fix

Changed Window.h to return the correct window handle. Also created an example function in gametechrenderer.cpp to demonstrate how ImGui works

KieranRodgerson committed:
Variable name change

Changed the windowHand variable I declared in Window.h to handle to bring it in line with other variables in the class.

KieranRodgerson committed:
Program now registers inputs to ImGui windows

KieranRodgerson committed:
Created UISystem class

Converted the code that was in GameTechRenderer into it's own class, UISystem to make the code cleaner.

KieranRodgerson committed:
Files Restructured

KieranRodgerson committed:
Framerate Counter Added

Added a DisplayFramerate function in UISystem and restructured UISystem to fix a bug

AlfieOnGit commented:
Unfortunately found a bug 😅 If you click the screen, it doesn't let you use your mouse to look around anymore, your view is stuck in place. Complete guess but maybe are you clicking some UI component so it's taking your focus off the main window logic?

Additionally, could the FPS count be updated less frequently so as to be more readable?
(e.g. maybe update every second and just showing to the nearest whole number. Could just literally have an int counter?)

KieranRodgerson committed:
FPS counter updates less frequently

FPS is now displayed less frequently than before to make it more readable

KieranRodgerson commented:
Sorry didn't see this comment. I've changed the FPS counter to update less frequently but I'm not getting the clicking screen bug on my end. Might have to take a look at what's happening on your machine when you're in the lab to try and figure out what's happening.

KieranRodgerson committed:
GameTimer fixed

GameTimer was causing build errors. Fixed these errors by including <chrono>

KieranRodgerson committed:
Merge branch 'main' into ui-impl

Alasdair-PB commented:
I'm having the same issue as Alfie with the PR. If I click on any UI widget the mouse no longer goes off the screen to allow the camera to scroll. I'm guessing something like `Window::GetWindow()->LockMouseToWindow(true);` is called when the window is focused.

Alasdair-PB commented:
Just looking through the PR. Can we move the contents around in this PR. I think the UIsystem* should probably be in TutorialGame. (TutorialGame will be heavily reworked at some point when I have a save system working, but it's good to get it in here for now).  Second can we move the imgui widget/base contents into a separate project (look at how Alfie did the Event project for reference with CMake) and make another project for our own UI systems: So the current UI demo and the UISystem.

KieranRodgerson committed:
Mouse lock bug fixed

KieranRodgerson committed:
ImGui library files moved to a seperate project

KieranRodgerson committed:
UI System moved to it's own project

KieranRodgerson commented:
I've fixed the bug that prevented the player from looking around after interacting with the UI. I've also moved the ImGui library files and the UISystem class to their own seperate projects. I've changed the cmakelists  so they should link up correctly. I could put UISystem* in tutorialgame but I'm not really sure what the purpose of it would be because it would need to be passed to the renderer anyway. ImGui has to be used as part of the render loop.

Alasdair-PB commented:
Is there a reason it needs to be the render loop? The renderer is updated in `TutorialGame` so it can be updated there easily.	`glEnable(GL_CULL_FACE);` functions can be called in your `UISystem` if they're needed for transparent UI panels. Code specific panels should not be in `GameTechRenderer` as `GameTechRenderer` should be completely independent from the scene/level loaded in. If it must be in `GameTechRenderer` then it should only have a UIManager.Render(); function where we add UI Render-ables to it in `TutorialGame`. (but I'd prefer if we didn't do this ,:)) )

KieranRodgerson committed:
Moved UISystem* to TutorialGame

The UISystem is now initialised in and can be called from TutorialGame. A DrawUIElements class has been added.

KieranRodgerson commented:
I've moved UISystem* into TutorialGame and created a function called DrawUIElements which is called in UpdateGame. UISystem::EndFrame still has to be called in the render loop in GameTechRenderer but it shouldn't need to be altered, it just renders the ui elements which have been drawn in TutorialGame. Becasuse of this, DrawUIElements has to be called before renderer->render in UpdateGame.

KieranRodgerson committed:
Merge branch 'main' into ui-impl

KieranRodgerson committed:
DrawDemo commented out

DrawDemo is now commented out, uncomment it if you want to see the demo ui window

KieranRodgerson committed:
Unnecessary variables removed from GameTechRenderer

Alasdair-PB reviewed (dismissed):
Looks good to me

AlfieOnGit reviewed (changes requested):
Yep, bug fixed, nice work.

I have just noticed something else though. In both the UISystem and ImGUI project CMakeLists.txt files, there's a link_library instruction pointing at the Events project. I'm guessing this is a typo but yh, the last line can be removed. I don't think it's actively causing problems since it's just copying the current code, but if I need to change around the Events project CMakeLists.txt file in the future, it could interfere with it (and lord knows CMake is already annoying enough to debug).

Also, while you're editing the CMake files, you could also `include_directories(../ImGui/)` in the System project. This would mean easier access to objects in ImGui project, so the `#include`s wouldn't have to be complete relative paths (i.e. `#include "imgui.h"` instead of `#include "../ImGui/imgui.h"`). Not a request, just an idea. It's why I included it in the Events project.

But yeah, once the Events bit is removed, should be good to go. I checked and it should still compile the same 😊

KieranRodgerson committed:
Changed CMake files

Removed link_library pointing towards events project

KieranRodgerson commented:
Yeah my bad, I had just copied the events cmakelists and modified it and forgot to remove the link_library instructions. Should be fixed now

KieranRodgerson committed:
Merge branch 'main' into ui-impl

mlhumphriss commented:
Tested current version, could interact with ui dropdown without issue

mlhumphriss reviewed (dismissed):
Passed testing, no errors

KieranRodgerson committed:
Merge branch 'main' into ui-impl

AlfieOnGit reviewed (approved):
Yep looks solved to me, perfect thanks!
